controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go

```diff
@@ -347,41 +324,117 @@ func (r *KubeadmControlPlaneReconciler) updateStatus(ctx context.Context, kcp *c
+func (r *KubeadmControlPlaneReconciler) upgradeControlPlane(ctx context.Context, cluster *clusterv1.Cluster, kcp *controlplanev1.KubeadmControlPlane, logger logr.Logger) (ctrl.Result, error) {
+	// TODO: handle reconciliation of etcd members and kubeadm config in case they get out of sync with cluster
+	ownedMachines, err := r.managementCluster.GetMachinesForCluster(ctx, clusterKey(cluster), internal.OwnedControlPlaneMachines(kcp.Name))
+	if err != nil {
+		logger.Error(err, "failed to retrieve machines for cluster")
+		return ctrl.Result{}, err
+	}
+	requireUpgrade := internal.UnionFilterMachines(
+		ownedMachines,
+		internal.HasOutdatedConfiguration(hash.Compute(&kcp.Spec)),
+		internal.OlderThan(kcp.Spec.UpgradeAfter),
+	)
+	if len(requireUpgrade) == 0 {
+		logger.Info("no machines found to upgrade")
+		// requeue to perform any additional operations that may be required
+		return ctrl.Result{Requeue: true}, nil
+	}
 
-	// TODO: verify health for each existing replica
-	// TODO: mark an old Machine via the label kubeadm.controlplane.cluster.x-k8s.io/selected-for-upgrade
-	// TODO: check full cluster health
-	// TODO: provision new Machine to replace marked Machine
-	// TODO: wait for health of all existing replicas, ideally in a re-entrant way to avoid waiting in process
-	// TODO: wait for full cluster health, ideally in a re-entrant way to avoid waiting in process
-	// TODO: Remove etcd membership for old Machine instance
-	// TODO: Update the kubeadm configmap
-	// TODO: Delete the Marked ControlPlane machine
-	// TODO: Continue with next OldMachine
+	if err := r.healthCheck(ctx, clusterKey(cluster), kcp.Name); err != nil {
+		logger.Error(err, "waiting for control plane to pass health check before continuing upgrade")
+		r.recorder.Eventf(kcp, corev1.EventTypeWarning, "ControlPlaneUnhealthy", "Waiting for control plane to pass health check before continuing upgrade: %v", err)
+		return ctrl.Result{RequeueAfter: HealthCheckFailedRequeueAfter}, err
+	}
 
-	return nil
+	// If there is not already a Machine that is marked for upgrade, find one and mark it
+	selectedForUpgrade := internal.FilterMachines(requireUpgrade, internal.SelectedForUpgrade())
+	if len(selectedForUpgrade) == 0 {
+		selectedMachine, err := r.selectMachineForUpgrade(ctx, cluster, requireUpgrade)
+		if err != nil {
+			logger.Error(err, "failed to select machine for upgrade")
+			return ctrl.Result{}, err
+		}
+		selectedForUpgrade = append(selectedForUpgrade, selectedMachine)
+	}
+
+	// Determine if we need to add an additional control plane Machine or clean up an already replaced one
+	switch {
+	// TODO: find a better comparison to use for adding a new ControlPlane instance. Since upgrades take precedence
+	// over scaling operations, we can't trust this to hold true
+	case len(ownedMachines) == int(*kcp.Spec.Replicas):
```

Comment:
This took me a second to understand, let me confirm my understanding is correct:

If we own the same number of machines as the number of requested replicas on the kubeadmcontrolplane, this indicates the control plane has enough machines. We can now scale the control plane up so we have n+1 owned machines when we've only asked for n replicas. Is this right?

Reply:
Assuming that prior to the upgrade all scaling operations have been completed. However we do not have any such assurances, so we need to find a better way to determine if we should be adding an additional control plane machine or deleting a machine that was marked for replacement.